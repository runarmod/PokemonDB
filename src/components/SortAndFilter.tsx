import { useState } from "react";
import SortFilter from "../assets/sort_filter.png";
import Star from "../assets/star_filled.png";
import X from "../assets/X.png";
import "../styles/SortAndFilter.css";
import { pokemonTypeColors, SortingType, useAppContext } from "../utils";
import PokemonTypeLabel from "./PokemonTypeLabel";

export default function SortAndFilter() {
    const { sortingOrder, changeSortingOrder, updateFilters, filters } =
        useAppContext();
    const [isMenuOpen, setIsMenuOpen] = useState(false);

    function handleRadioChange(event: React.ChangeEvent<HTMLInputElement>) {
        changeSortingOrder(Number(event.target.value));
    }

    function handleCheckChange(event: React.ChangeEvent<HTMLInputElement>) {
        updateFilters(event.target.value);
    }

    return (
        <div id="SortAndFilterWrapper">
            <button
                onClick={() => setIsMenuOpen(!isMenuOpen)}
                className={`sortFilterMenu ${isMenuOpen ? "active" : ""}`}
                aria-label={
                    isMenuOpen ? "Close menu" : "Open sort/filter options"
                }
            >
                <img
                    src={isMenuOpen ? X : SortFilter}
                    alt={isMenuOpen ? "Close" : "Sort/filter"}
                />
            </button>
            <div className={`sortAndFilter ${isMenuOpen ? "active" : ""}`}>
                <h2>Sort by</h2>
                <ul className="sortList">
                    <li className="sortListEntry">
                        <input
                            type="radio"
                            name="sorting"
                            id="rbtId"
                            value={SortingType.ID}
                            onChange={handleRadioChange}
                            checked={sortingOrder == SortingType.ID}
                        />
                        <label htmlFor="rbtId">#ID</label>
                    </li>
                    <li className="sortListEntry">
                        <input
                            type="radio"
                            name="sorting"
                            id="rbtName"
                            value={SortingType.NAME_ASC}
                            onChange={handleRadioChange}
                            checked={sortingOrder == SortingType.NAME_ASC}
                        />
                        <label htmlFor="rbtName">Name A-Z</label>
                    </li>
                    <li className="sortListEntry">
                        <input
                            type="radio"
                            name="sorting"
                            id="rbtNameReversed"
                            value={SortingType.NAME_DESC}
                            onChange={handleRadioChange}
                            checked={sortingOrder == SortingType.NAME_DESC}
                        />
                        <label htmlFor="rbtNameReversed">Name Z-A</label>
                    </li>
                </ul>
                <h2>Filters</h2>
                <ul className="filterList">
                    <li className="filterListEntry">
                        <input
                            type="checkbox"
                            id="checkFavorite"
                            value={"favorite"}
                            onChange={handleCheckChange}
                            checked={filters.includes("favorite")}
                        />
                        <label htmlFor="checkFavorite">
                            <img
                                src={Star}
                                alt="favorite icon"
                                id="favoriteIcon"
                            />
                        </label>
                    </li>

                    {Array.from(pokemonTypeColors.keys()).map(
                        (type: string) => (
                            <li key={type} className="filterListEntry">
                                <input
                                    type="checkbox"
                                    id={"check" + type}
                                    value={type}
                                    onChange={handleCheckChange}
                                    checked={filters.includes(type)}
                                />
                                <label htmlFor={"check" + type}>
                                    <PokemonTypeLabel type={type} />
                                </label>
                            </li>
                        )
                    )}
                </ul>
            </div>
        </div>
    );
}
