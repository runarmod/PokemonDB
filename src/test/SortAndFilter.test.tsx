import { describe, it, expect, vi, Mock } from "vitest";
import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { SortingType, useAppContext } from "../utils";
import "@testing-library/jest-dom";
import SortAndFilter from "../components/SortAndFilter";

vi.mock(import("../utils"), async (importOriginal) => {
    const actual = await importOriginal();
    return {
        ...actual,
        useAppContext: vi.fn(),
    };
});

describe("SortAndFilter - General tests", () => {
    describe("Sorting Options", () => {
        it("renders sorting options", () => {
            (useAppContext as Mock).mockReturnValue({
                sortingOrder: SortingType.ID,
                changeSortingOrder: vi.fn(),
                updateFilters: vi.fn(),
                filters: [],
            });
            render(<SortAndFilter />);

            expect(screen.getByLabelText("#ID")).toBeInTheDocument();
            expect(screen.getByLabelText("Name A-Z")).toBeInTheDocument();
            expect(screen.getByLabelText("Name Z-A")).toBeInTheDocument();

            const radioButtons = screen.getAllByRole("radio");
            expect(radioButtons).toHaveLength(3);
        });

        it("renders ID sorting option", () => {
            (useAppContext as Mock).mockReturnValue({
                sortingOrder: SortingType.ID,
                changeSortingOrder: vi.fn(),
                updateFilters: vi.fn(),
                filters: [],
            });
            render(<SortAndFilter />);

            const checkBox = screen.getByLabelText("#ID");
            expect(checkBox).toBeInTheDocument();
            expect(checkBox).toBeChecked();

            const radioButtons = screen.getAllByRole("radio");
            const selectedRadioButtons = radioButtons.filter(
                (radioButton) => (radioButton as HTMLInputElement).checked
            );
            expect(selectedRadioButtons).toHaveLength(1);
        });

        it("renders ascending sorting option", () => {
            (useAppContext as Mock).mockReturnValue({
                sortingOrder: SortingType.NAME_ASC,
                changeSortingOrder: vi.fn(),
                updateFilters: vi.fn(),
                filters: [],
            });
            render(<SortAndFilter />);

            const checkBox = screen.getByLabelText("Name A-Z");
            expect(checkBox).toBeInTheDocument();
            expect(checkBox).toBeChecked();

            const radioButtons = screen.getAllByRole("radio");
            const selectedRadioButtons = radioButtons.filter(
                (radioButton) => (radioButton as HTMLInputElement).checked
            );
            expect(selectedRadioButtons).toHaveLength(1);
        });

        it("renders descending sorting option", () => {
            (useAppContext as Mock).mockReturnValue({
                sortingOrder: SortingType.NAME_DESC,
                changeSortingOrder: vi.fn(),
                updateFilters: vi.fn(),
                filters: [],
            });
            render(<SortAndFilter />);

            const checkBox = screen.getByLabelText("Name Z-A");
            expect(checkBox).toBeInTheDocument();
            expect(checkBox).toBeChecked();

            const radioButtons = screen.getAllByRole("radio");
            const selectedRadioButtons = radioButtons.filter(
                (radioButton) => (radioButton as HTMLInputElement).checked
            );
            expect(selectedRadioButtons).toHaveLength(1);
        });

        it("should not change sortingOrder when clicking the same radiobutton", async () => {
            const mockChangeSortingOrder = vi.fn();
            (useAppContext as Mock).mockReturnValue({
                sortingOrder: SortingType.ID,
                changeSortingOrder: mockChangeSortingOrder,
                updateFilters: vi.fn(),
                filters: ["favorite", "fire", "water"],
            });
            render(<SortAndFilter />);

            const radioButton = screen.getByLabelText("#ID");
            expect(radioButton).toBeInTheDocument();
            expect(radioButton).toBeChecked();
            userEvent.click(radioButton);

            await waitFor(() => {
                expect(mockChangeSortingOrder).toBeCalledTimes(0);
            });
        });

        it("should change when clicking another radiobutton", async () => {
            const mockChangeSortingOrder = vi.fn();
            (useAppContext as Mock).mockReturnValue({
                sortingOrder: SortingType.ID,
                changeSortingOrder: mockChangeSortingOrder,
                updateFilters: vi.fn(),
                filters: ["favorite", "fire", "water"],
            });
            render(<SortAndFilter />);

            const radioButton = screen.getByLabelText("#ID");
            expect(radioButton).toBeInTheDocument();
            expect(radioButton).toBeChecked();
            const radioButtonASC = screen.getByLabelText("Name A-Z");
            userEvent.click(radioButtonASC);

            await waitFor(() => {
                expect(mockChangeSortingOrder).toHaveBeenCalledTimes(1);
                expect(mockChangeSortingOrder).toHaveBeenCalledWith(
                    SortingType.NAME_ASC
                );
            });
        });
    });

    describe("Filters", () => {
        it("renders filters", () => {
            (useAppContext as Mock).mockReturnValue({
                sortingOrder: SortingType.ID,
                changeSortingOrder: vi.fn(),
                updateFilters: vi.fn(),
                filters: [],
            });
            render(<SortAndFilter />);

            const filledStarIcon = screen.getByAltText("favorite icon");
            expect(filledStarIcon).toBeInTheDocument();
            expect(filledStarIcon).toHaveAttribute(
                "src",
                "/src/assets/star_filled.png"
            );
            expect(screen.getByLabelText("normal")).toBeInTheDocument();
            expect(screen.getByLabelText("fire")).toBeInTheDocument();
            expect(screen.getByLabelText("water")).toBeInTheDocument();
            expect(screen.getByLabelText("electric")).toBeInTheDocument();
            expect(screen.getByLabelText("grass")).toBeInTheDocument();
            expect(screen.getByLabelText("ice")).toBeInTheDocument();
            expect(screen.getByLabelText("fighting")).toBeInTheDocument();
            expect(screen.getByLabelText("poison")).toBeInTheDocument();
            expect(screen.getByLabelText("ground")).toBeInTheDocument();
            expect(screen.getByLabelText("flying")).toBeInTheDocument();
            expect(screen.getByLabelText("psychic")).toBeInTheDocument();
            expect(screen.getByLabelText("bug")).toBeInTheDocument();
            expect(screen.getByLabelText("ghost")).toBeInTheDocument();
            expect(screen.getByLabelText("dragon")).toBeInTheDocument();
            expect(screen.getByLabelText("dark")).toBeInTheDocument();
            expect(screen.getByLabelText("steel")).toBeInTheDocument();
            expect(screen.getByLabelText("fairy")).toBeInTheDocument();
            expect(screen.getByLabelText("rock")).toBeInTheDocument();

            const checkBoxes = screen.getAllByRole("checkbox");
            expect(checkBoxes).toHaveLength(19); // 18 types + favorites

            const checkedCheckboxes = checkBoxes.filter(
                (checkBox) => (checkBox as HTMLInputElement).checked
            );
            expect(checkedCheckboxes).toHaveLength(0);
        });

        it("renders correct filters", () => {
            (useAppContext as Mock).mockReturnValue({
                sortingOrder: SortingType.ID,
                changeSortingOrder: vi.fn(),
                updateFilters: vi.fn(),
                filters: ["favorite", "fire", "water"],
            });
            render(<SortAndFilter />);

            const favoriteCheckbox = screen.getByRole("checkbox", {
                name: "favorite icon",
            });
            const fireCheckbox = screen.getByRole("checkbox", { name: "fire" });
            const waterCheckbox = screen.getByRole("checkbox", {
                name: "water",
            });
            expect(favoriteCheckbox).toBeChecked();
            expect(fireCheckbox).toBeChecked();
            expect(waterCheckbox).toBeChecked();

            const checkBoxes = screen.getAllByRole("checkbox");
            const checkedCheckboxes = checkBoxes.filter(
                (checkBox) => (checkBox as HTMLInputElement).checked
            );
            expect(checkedCheckboxes).toHaveLength(3);
        });

        it("should add filter when clicking unchecked checkbox", async () => {
            const mockUpdateFilters = vi.fn();
            (useAppContext as Mock).mockReturnValue({
                sortingOrder: SortingType.ID,
                changeSortingOrder: vi.fn(),
                updateFilters: mockUpdateFilters,
                filters: ["favorite", "fire"],
            });
            render(<SortAndFilter />);

            const waterCheckbox = screen.getByRole("checkbox", {
                name: "water",
            });
            expect(waterCheckbox).toBeInTheDocument();
            expect((waterCheckbox as HTMLInputElement).checked).toBe(false);
            userEvent.click(waterCheckbox);
            await waitFor(() => {
                expect(mockUpdateFilters).toHaveBeenCalledTimes(1);
                expect(mockUpdateFilters).toHaveBeenCalledWith("water");
            });
        });

        it("should remove filter when clicking checked checkbox", async () => {
            const mockUpdateFilters = vi.fn();
            (useAppContext as Mock).mockReturnValue({
                sortingOrder: SortingType.ID,
                changeSortingOrder: vi.fn(),
                updateFilters: mockUpdateFilters,
                filters: ["favorite", "fire", "water"],
            });
            render(<SortAndFilter />);

            const fireCheckbox = screen.getByRole("checkbox", { name: "fire" });
            expect(fireCheckbox).toBeInTheDocument();
            expect(fireCheckbox).toBeChecked();
            userEvent.click(fireCheckbox);
            await waitFor(() => {
                expect(mockUpdateFilters).toHaveBeenCalledTimes(1);
                expect(mockUpdateFilters).toHaveBeenCalledWith("fire");
            });
        });
    });

    describe("isMenuOpen test", () => {
        it("should change state (and alt-text as a result) when clicking on menu button", async () => {
            (useAppContext as Mock).mockReturnValue({
                sortingOrder: SortingType.ID,
                changeSortingOrder: vi.fn(),
                updateFilters: vi.fn(),
                filters: [],
            });
            render(<SortAndFilter />);
            const menuButton = screen.getByAltText("Sort/filter");
            expect(menuButton).toBeInTheDocument();
            userEvent.click(menuButton);

            await waitFor(() => {
                expect(screen.getByAltText("Close")).toBeInTheDocument();
            });
        });
    });

    describe("Snapshot test", () => {
        it("matches snapshot", () => {
            (useAppContext as Mock).mockReturnValue({
                sortingOrder: SortingType.ID,
                changeSortingOrder: vi.fn(),
                updateFilters: vi.fn(),
                filters: ["favorite", "fire", "water"],
            });

            const { asFragment } = render(<SortAndFilter />);
            expect(asFragment()).toMatchSnapshot();
        });
    });
});
