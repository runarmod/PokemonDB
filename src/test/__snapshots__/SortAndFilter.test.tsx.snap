// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SortAndFilter - General tests > Snapshot test > matches snapshot 1`] = `
<DocumentFragment>
  <div
    id="SortAndFilterWrapper"
  >
    <button
      aria-label="Open sort/filter options"
      class="sortFilterMenu "
    >
      <img
        alt="Sort/filter"
        src="/src/assets/sort_filter.png"
      />
    </button>
    <div
      class="sortAndFilter "
    >
      <h2>
        Sort by
      </h2>
      <ul
        class="sortList"
      >
        <li
          class="sortListEntry"
        >
          <input
            checked=""
            id="rbtId"
            name="sorting"
            type="radio"
            value="2"
          />
          <label
            for="rbtId"
          >
            #ID
          </label>
        </li>
        <li
          class="sortListEntry"
        >
          <input
            id="rbtName"
            name="sorting"
            type="radio"
            value="0"
          />
          <label
            for="rbtName"
          >
            Name A-Z
          </label>
        </li>
        <li
          class="sortListEntry"
        >
          <input
            id="rbtNameReversed"
            name="sorting"
            type="radio"
            value="1"
          />
          <label
            for="rbtNameReversed"
          >
            Name Z-A
          </label>
        </li>
      </ul>
      <h2>
        Filters
      </h2>
      <ul
        class="filterList"
      >
        <li
          class="filterListEntry"
        >
          <input
            checked=""
            id="checkFavorite"
            type="checkbox"
            value="favorite"
          />
          <label
            for="checkFavorite"
          >
            <img
              alt="favorite icon"
              id="favoriteIcon"
              src="/src/assets/star_filled.png"
            />
          </label>
        </li>
        <li
          class="filterListEntry"
        >
          <input
            id="checknormal"
            type="checkbox"
            value="normal"
          />
          <label
            for="checknormal"
          >
            <span
              class="pokemonTypeLabel"
              style="background-color: rgb(159, 161, 159);"
            >
              normal
            </span>
          </label>
        </li>
        <li
          class="filterListEntry"
        >
          <input
            checked=""
            id="checkfire"
            type="checkbox"
            value="fire"
          />
          <label
            for="checkfire"
          >
            <span
              class="pokemonTypeLabel"
              style="background-color: rgb(230, 40, 41);"
            >
              fire
            </span>
          </label>
        </li>
        <li
          class="filterListEntry"
        >
          <input
            checked=""
            id="checkwater"
            type="checkbox"
            value="water"
          />
          <label
            for="checkwater"
          >
            <span
              class="pokemonTypeLabel"
              style="background-color: rgb(41, 128, 239);"
            >
              water
            </span>
          </label>
        </li>
        <li
          class="filterListEntry"
        >
          <input
            id="checkelectric"
            type="checkbox"
            value="electric"
          />
          <label
            for="checkelectric"
          >
            <span
              class="pokemonTypeLabel"
              style="background-color: rgb(250, 192, 0);"
            >
              electric
            </span>
          </label>
        </li>
        <li
          class="filterListEntry"
        >
          <input
            id="checkgrass"
            type="checkbox"
            value="grass"
          />
          <label
            for="checkgrass"
          >
            <span
              class="pokemonTypeLabel"
              style="background-color: rgb(63, 161, 41);"
            >
              grass
            </span>
          </label>
        </li>
        <li
          class="filterListEntry"
        >
          <input
            id="checkice"
            type="checkbox"
            value="ice"
          />
          <label
            for="checkice"
          >
            <span
              class="pokemonTypeLabel"
              style="background-color: rgb(61, 206, 243);"
            >
              ice
            </span>
          </label>
        </li>
        <li
          class="filterListEntry"
        >
          <input
            id="checkfighting"
            type="checkbox"
            value="fighting"
          />
          <label
            for="checkfighting"
          >
            <span
              class="pokemonTypeLabel"
              style="background-color: rgb(255, 128, 0);"
            >
              fighting
            </span>
          </label>
        </li>
        <li
          class="filterListEntry"
        >
          <input
            id="checkpoison"
            type="checkbox"
            value="poison"
          />
          <label
            for="checkpoison"
          >
            <span
              class="pokemonTypeLabel"
              style="background-color: rgb(145, 65, 203);"
            >
              poison
            </span>
          </label>
        </li>
        <li
          class="filterListEntry"
        >
          <input
            id="checkground"
            type="checkbox"
            value="ground"
          />
          <label
            for="checkground"
          >
            <span
              class="pokemonTypeLabel"
              style="background-color: rgb(145, 81, 33);"
            >
              ground
            </span>
          </label>
        </li>
        <li
          class="filterListEntry"
        >
          <input
            id="checkflying"
            type="checkbox"
            value="flying"
          />
          <label
            for="checkflying"
          >
            <span
              class="pokemonTypeLabel"
              style="background-color: rgb(129, 185, 239);"
            >
              flying
            </span>
          </label>
        </li>
        <li
          class="filterListEntry"
        >
          <input
            id="checkpsychic"
            type="checkbox"
            value="psychic"
          />
          <label
            for="checkpsychic"
          >
            <span
              class="pokemonTypeLabel"
              style="background-color: rgb(239, 65, 121);"
            >
              psychic
            </span>
          </label>
        </li>
        <li
          class="filterListEntry"
        >
          <input
            id="checkbug"
            type="checkbox"
            value="bug"
          />
          <label
            for="checkbug"
          >
            <span
              class="pokemonTypeLabel"
              style="background-color: rgb(145, 161, 25);"
            >
              bug
            </span>
          </label>
        </li>
        <li
          class="filterListEntry"
        >
          <input
            id="checkghost"
            type="checkbox"
            value="ghost"
          />
          <label
            for="checkghost"
          >
            <span
              class="pokemonTypeLabel"
              style="background-color: rgb(112, 65, 112);"
            >
              ghost
            </span>
          </label>
        </li>
        <li
          class="filterListEntry"
        >
          <input
            id="checkdragon"
            type="checkbox"
            value="dragon"
          />
          <label
            for="checkdragon"
          >
            <span
              class="pokemonTypeLabel"
              style="background-color: rgb(80, 96, 225);"
            >
              dragon
            </span>
          </label>
        </li>
        <li
          class="filterListEntry"
        >
          <input
            id="checkdark"
            type="checkbox"
            value="dark"
          />
          <label
            for="checkdark"
          >
            <span
              class="pokemonTypeLabel"
              style="background-color: rgb(98, 77, 78);"
            >
              dark
            </span>
          </label>
        </li>
        <li
          class="filterListEntry"
        >
          <input
            id="checksteel"
            type="checkbox"
            value="steel"
          />
          <label
            for="checksteel"
          >
            <span
              class="pokemonTypeLabel"
              style="background-color: rgb(96, 161, 184);"
            >
              steel
            </span>
          </label>
        </li>
        <li
          class="filterListEntry"
        >
          <input
            id="checkfairy"
            type="checkbox"
            value="fairy"
          />
          <label
            for="checkfairy"
          >
            <span
              class="pokemonTypeLabel"
              style="background-color: rgb(239, 112, 239);"
            >
              fairy
            </span>
          </label>
        </li>
        <li
          class="filterListEntry"
        >
          <input
            id="checkrock"
            type="checkbox"
            value="rock"
          />
          <label
            for="checkrock"
          >
            <span
              class="pokemonTypeLabel"
              style="background-color: rgb(175, 169, 129);"
            >
              rock
            </span>
          </label>
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;
